libbpf-tools/offcputime: Add -P option to post unwind user stack backtrace using dwarf unwind via libunwind

ekyooo opened the pull request:
This is not the finished version for review. Before I finish, I'd like to know that this feature is required by bcc.
Can you please let me know if post unwind using dwarf unwind via libunwind is needed and it helps with issues like #3515?

FYI, there are a few things left to do for this PR if needed:
- Added ARCH dependent code for x86
- Dump user stack and regs when in kernel mode
- Add libunwind for build
- etc.

1. test environment
Tested with profile in aarch64 environment.
However, the profile's review PR is not in progress, so I apply it to offcputime to share the code.

2. This is the test results.
Before: 
```
# ./profile -d -p `pidof test-strlen-abc-nofp-unwtbl`
Sampling at 49 Hertz of PID 83030 by user + kernel stack... Hit Ctrl-C to end.
^C    
    #0  0x0000007fbe74cc20 strerrordesc_np+0x1980 (/usr/lib/libc.so.6+0x9cc20)
    #1  0x0000007fbe6db30c __libc_start_main+0x9c (/usr/lib/libc.so.6+0x2b30c)
    #2  0x000000556f9c0754 _start+0x34 (/var/rootdirs/home/root/test-strlen-abc-nofp-unwtbl+0x754)
    -                test-strlen-abc (83030)
        290
```

After:
```
# ./profile -P -d -p `pidof test-strlen-abc-nofp-unwtbl`
Sampling at 49 Hertz of PID 83030 by user + kernel stack... Hit Ctrl-C to end.
^C    
    #0  0x0000007fbe74cc20 strerrordesc_np+0x1980 (/usr/lib/libc.so.6+0x9cc20)
    #1  0x000000556f9c087c c+0x30 (/var/rootdirs/home/root/test-strlen-abc-nofp-unwtbl+0x87c)
    #2  0x000000556f9c0840 b+0x8 (/var/rootdirs/home/root/test-strlen-abc-nofp-unwtbl+0x840)
    #3  0x000000556f9c082c a+0x8 (/var/rootdirs/home/root/test-strlen-abc-nofp-unwtbl+0x82c)
    #4  0x000000556f9c08c4 main+0x10 (/var/rootdirs/home/root/test-strlen-abc-nofp-unwtbl+0x8c4)
    #6  0x0000007fbe6db230 __libc_init_first+0x70 (/usr/lib/libc.so.6+0x2b230)
    -                test-strlen-abc (83030)
        1
```
3. test executable information
Test executable has no frame pointer, but has an unwind table (.eh_frame).
3.1 compile option : -fomit-frame-pointer -funwind-tables 
3.2 code 
```
#include <assert.h>
#include <stdlib.h>
#include <string.h>

char buf[8*1024*1024];

int a(void);
int b(void);
int c(void);

typedef size_t (*strlen_t)(const char *); 
strlen_t getlen = strlen;

int a(void)
{
        return b() - 1;
}

int b(void)
{
        return c() + 1;
}

int c() 
{
        memset(buf, '+', sizeof(buf)-1);

        while (1) {
                size_t len = getlen(buf);
                assert(len == sizeof(buf) - 1); 
        }

        // not reached
        return 1;
}

int main(int argc, char **argv)
{
        a();
}
```

Thank you.

yonghong-song commented:
@davemarchevsky could you also take a look at this patch?

yonghong-song commented:
I think this could be a useful feature. In most bcc tools, the user space gets stack_id and a list of addresses for the stack and then reconstruct the symbolized stack strace based recorded pid. If the application compiled with omitting frame pointer, it would be hard from the kernel to figure out the correct stack trace. In these cases, having a backup using user space libunwind seems a reasonable alternative. So please go ahead to finish your implementation. I see you use stack size 128. This may not be enough for real world big application which could be up to 4k. I think we should make stack size as part of the input for -P. If not specified, you could use 128 bytes.

ekyooo commented:
Thank you for your answers and comments.
I'll go ahead to finish the implementation. 
Thank you.

ekyooo committed:
libbpf-tools/unwind_helpers: Add helper API for post dwarf unwind using libunwind

unwind_helpers.c/.h:
  User program helper API for post-DWARF unwinding using libunwind
unwind_helpers.bpf.h:
  BPF program helper providing API to return sample (user stack and register) IDs

ekyooo committed:
libbpf-tools/offcputime: Add -D option for post dwarf unwind using unwind_helpers

Add -D option to use post dwarf unwind with libunwind.
With the option, bpf program dump user stack and user regs.
and user program get user stack traces using unwind_helpers to show output.

ekyooo commented:
Not yet subject to review. Thank you.

ekyooo committed:
libbpf-tools/unwind_helpers: Add build configuration to Makefile

Use installed libunwind

Signed-off-by: Eunseon Lee <es.lee@lge.com>

ekyooo commented on libbpf-tools/Makefile:
```diff
@@ -26,6 +29,10 @@ USE_BLAZESYM ?= 1
 endif
 
+ifeq ($(ARCH),x86)
+USE_LIBUNWIND = 1
```
Could you please tell me if LIBUNWIND should be enabled by default?

ekyooo commented on libbpf-tools/offcputime.c:
```diff
@@ -154,6 +161,23 @@ static error_t parse_arg(int key, char *arg, struct argp_state *state)
+	case 'P':
+		env.unwind = true;
+		break;
+	case OPT_SAMPLE_STACK_SIZE:
```
> I think we should make stack size as part of the input for -P. If not specified, you could use 128 bytes.

offcputime has a positional argument for the duration, so -P cannot be set to OPTION_ARG_OPTIONAL.
Therefore, a separate option for stack size has been added.

ekyooo commented on libbpf-tools/Makefile:
```diff
@@ -26,6 +29,10 @@ USE_BLAZESYM ?= 1
 endif
 endif
 
+ifeq ($(ARCH),x86)
```
For now, only build libunwind for x86 is added.

ekyooo replied:
This is a test example using the test application in the description.

```
# ./offcputime -P --sample-stack-size 1024 -p `pidof test-strlen-abc-nofp-unwtbl` 
^C    
    bpf_prog_99b48c45746a8a74_sched_switch 
    bpf_prog_99b48c45746a8a74_sched_switch   
    bpf_trace_run3                           
    __traceiter_sched_switch                 
    __schedule                               
    schedule                                 
    exit_to_user_mode_prepare                
    irqentry_exit_to_user_mode               
    asm_sysvec_apic_timer_interrupt          
    strlen                                   
    c                                        
    b                                        
    a                                        
    main                                     
    __libc_start_main                        
    _start                                   
    [unknown]                                
    [unknown]                                
    -                test-strlen-abc (26307) 
        50                                   
```
test with -v option.
```
# ./offcputime -P -v --sample-stack-size 1024 -p `pidof test-strlen-abc-nofp-unwtbl` 
    #0  0xffffffffc001259c bpf_prog_99b48c45746a8a74_sched_switch+0x4bc         
    #1  0xffffffffc001259c bpf_prog_99b48c45746a8a74_sched_switch+0x4bc         
    #2  0xffffffff8116952e bpf_trace_run3+0x3e                                  
    #3  0xffffffff8109be54 __traceiter_sched_switch+0x34                        
    #4  0xffffffff81c5f7ba __schedule+0x30a                                     
    #5  0xffffffff81c5fbff schedule+0x3f                                        
    #6  0xffffffff810f3766 exit_to_user_mode_prepare+0x56                       
    #7  0xffffffff81c5d6d5 irqentry_exit_to_user_mode+0x5                       
    #8  0xffffffff81e00c42 asm_sysvec_apic_timer_interrupt+0x12                 
    #9  0x00007f9797d6fbc6 strlen+0x106 (/lib/x86_64-linux-gnu/libc-2.19.so+0x88bc6)                                                                            
    #10 0x00005582b6600716 c+0x2a (/root/es/unwind/test-strlen-abc-nofp-unwtbl+0x716)                                                                           
    #11 0x00005582b66006e4 b+0x9 (/root/es/unwind/test-strlen-abc-nofp-unwtbl+0x6e4)                                                                            
    #12 0x00005582b66006d3 a+0x9 (/root/es/unwind/test-strlen-abc-nofp-unwtbl+0x6d3)                                                                            
    #13 0x00005582b6600756 main+0x11 (/root/es/unwind/test-strlen-abc-nofp-unwtbl+0x756)                                                                        
    #14 0x00007f9797d08ec5 __libc_start_main+0xf5 (/lib/x86_64-linux-gnu/libc-2.19.so+0x21ec5)                                                                  
    #15 0x00005582b66005ea _start+0x2a (/root/es/unwind/test-strlen-abc-nofp-unwtbl+0x5ea)                                                                      
    #16 0xffffffff81c5d6d5 (/root/es/unwind/test-strlen-abc-nofp-unwtbl+0x5ea)  
    #17 0xffffffff81e00c42 (/root/es/unwind/test-strlen-abc-nofp-unwtbl+0x5ea)  
    -                test-strlen-abc (26307)                                    
        54                                                                      
 ```

Thank you for waiting so long.
Could you please review this?
Thank you.

ekyooo commented:
In the last update, this feature is only built if USE_LIBUNWIND is enabled in the Makefile.
Thank you.

Bojun-Seo commented on libbpf-tools/offcputime.c:
```diff
@@ -355,9 +427,25 @@ int main(int argc, char **argv)
 	 */
 	sleep(env.duration);
 
+#ifdef USE_LIBUNWIND
```
How about removing `USE_LIBUNWIND` macros?

Bojun-Seo commented on libbpf-tools/offcputime.c:
```diff
@@ -239,10 +288,22 @@ static void print_map(struct ksyms *ksyms, struct syms_cache *syms_cache,
+#ifdef USE_LIBUNWIND
+		if (env.unwind) {
+			if (post_unwind(u, sample_fd, ustack_fd, next_key.user_stack_id,
+					ip, env.perf_max_stack_depth) != 0) {
```
It would be nice if `ip` were the last parameter.

ekyooo commented:
@Bojun-Seo 
Thank you for your comment.
Including your feedback, I've found many things I need to fix and improve in this patch.
Well, I'll come back soon with a revised patch.
Thanks again.

ekyooo committed:
libbpf-tools/memleak: Add -D option for post dwarf unwind using unwind_helpers

Add -D option to use post dwarf unwind with libunwind.
With the option, bpf program dump user stack and user regs.
and user program get user stack traces using unwind_helpers to show output.

ekyooo commented:
Apply the post unwind feature to the memleak tool for testing in relation to issue https://github.com/iovisor/bcc/issues/4958.
Not ready for review yet. 
Thank you.

**Test result**

*Before*

```
root@ubuntu-Standard-PC-i440FX-PIIX-1996:~/es/memleak# ./memleak -p `pidof leak_gen`                                                                             
using default object: libc.so.6                                                 
using page size: 4096                                                           
tracing kernel: false                                                           
Tracing outstanding memory allocs...  Hit Ctrl-C to end                         
^C[5:22:11] Top 1 stacks with outstanding allocations:                          
4 bytes in 1 allocations from stack                                             
        0 [<00007f5747c69dad>] _Znwm+0x1d [/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.19]                                                                       
        1 [<000055a6698007e3>] main+0x49 [/root/es/unwind/leak_gen]              
        2 [<00007f5747867ec5>] __libc_start_main+0xf5 [/lib/x86_64-linux-gnu/libc-2.19.so]                                                                      
done      
```

*After*

with Post unwinding (-U option in memleak)

```
root@ubuntu-Standard-PC-i440FX-PIIX-1996:~/es/memleak# ./memleak -U -p `pidof leak_gen`                                                                          
using default object: libc.so.6                                                 
using page size: 4096                                                           
tracing kernel: false                                                           
libbpf: Error in bpf_create_map_xattr(uw_stacks):Invalid argument(-22). Retrying without BTF.                                                                   
Tracing outstanding memory allocs...  Hit Ctrl-C to end                         
^C[5:23:11] Top 1 stacks with outstanding allocations:                          
4 bytes in 1 allocations from stack                                             
        0 [<00007f5747c69dad>] _Znwm+0x1d [/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.19]                                                                       
        1 [<000055a669800798>] _Z3foov+0xe [/root/es/unwind/leak_gen]            
        2 [<000055a6698007de>] main+0x44 [/root/es/unwind/leak_gen]              
        3 [<00007f5747867ec5>] __libc_start_main+0xf5 [/lib/x86_64-linux-gnu/libc-2.19.so]                                                                      
        4 [<000055a6698006aa>] _start+0x2a [/root/es/unwind/leak_gen]            
done     
```

ekyooo commented:
Unified the option names for offcputime and memleak.
```
# ./memleak --help
  -D, --dwarf-unwind         Enable DWARF mode stack traces
      --dwarf-stack-size=STACK-SIZE
                             Set user stack size for DWARF mode (default 128)
```

Not ready for review yet.
Thank you.

**Test result**

```
# ./memleak_dbg -D --dwarf-stack-size 512 -p `pidof leak_gen`
Tracing outstanding memory allocs...  Hit Ctrl-C to end
[7:36:35] Top 1 stacks with outstanding allocations:
4 bytes in 1 allocations from stack
	0 [<000071621782798c>] _Znwm+0x1c [/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30]
	1 [<00005c9469c007d2>] _Z3foov+0x48 [/workspace/temp/leak_gen]
	2 [<00005c9469c007fd>] main+0x11 [/workspace/temp/leak_gen]
	3 [<000071621757ad90>] [/usr/lib/x86_64-linux-gnu/libc.so.6]
	4 [<000071621757ae40>] __libc_start_main+0x80 [/usr/lib/x86_64-linux-gnu/libc.so.6]
	5 [<00005c9469c006aa>] _start+0x2a [/workspace/temp/leak_gen]
done
```

ekyooo commented:
**Changes:**
- memleak: the maximum number of stack entries for DWARF mode can be changed using the --stack-storage-size option.
- memleak: applying several patches to address stack max entry overflow situations.
  - https://github.com/iovisor/bcc/pull/5237
  - https://github.com/iovisor/bcc/pull/5238
  - https://github.com/iovisor/bcc/pull/5240 
  - The memleak patches will be excluded from the commit list of this PR once they are merged into the master branch.
- unwind_helpers: return error code for stack max entry overflow.



**Usage example:**
- Increase the maximum number of stack entries
```
# ./memleak -D --dwarf-stack-size 256 --stack-storage-size 102400 -p `pidof t3` 
```

- warning message for stack entries overflow
```
root@58a84cfc328d:/workspace/temp# ./memleak -D --stack-storage-size 1 -p `pidof t3` 
^C[11:57:52] Top 1 stacks with outstanding allocations:
4 bytes in 1 allocations from stack
	0 [<00007afda6f9198c>] _Znwm+0x1c [/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.30]
	1 [<0000613cd73871f5>] _Z3foov+0x4c [/workspace/temp/t3]
	2 [<0000613cd7387224>] main+0x15 [/workspace/temp/t3]
	3 [<00007afda6ce4d90>] [/usr/lib/x86_64-linux-gnu/libc.so.6]
WARNING: 3 stack traces could not be displayed.  // <------------------- need to use  --stack-storage-size option
done
```

ekyooo commented:
**Changes:**
The memleak improvement patches are excluded from this PR's commit list as they have already been merged into the master branch.

Not ready for review yet.
Thank you.

ekyooo commented:
> Hi ekyooo, Based on your branch, I tried running offcpu locally. After enabling the unwind feature, user-space stacks are not showing up. Could this be an issue with my build/compile environment? @ekyooo

- **Cause**: When the buffer passed to unwind_helper for reading stacktrace in offcputime contains garbage values, they get printed in the output.
- **Fix**: Added memset inside unwind_helper to clear the caller-provided buffer, similar to how the kernel stackmap works. (Though this doesn't seem like an ideal design to me)
- Note: The original behavior of unwind_helper was to return the number of IPs written to the buffer without doing memset internally.

Could you check with this version? 

Thank you.

chentao-kernel commented:
> > Hi ekyooo, Based on your branch, I tried running offcpu locally. After enabling the unwind feature, user-space stacks are not showing up. Could this be an issue with my build/compile environment? @ekyooo
> 
> * **Cause**: When the buffer passed to unwind_helper for reading stacktrace in offcputime contains garbage values, they get printed in the output.
> * **Fix**: Added memset inside unwind_helper to clear the caller-provided buffer, similar to how the kernel stackmap works. (Though this doesn't seem like an ideal design to me)
> * Note: The original behavior of unwind_helper was to return the number of IPs written to the buffer without doing memset internally.
> 
> Could you check with this version?
> 
> Thank you.

I try it, thanks
